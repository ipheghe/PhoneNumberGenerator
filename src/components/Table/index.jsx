// react Lib imports
import React from 'react';

// third-party imoports
import PropTypes from 'prop-types';

// styles
import './Table.scss';

/**
 * This component displays a table with props data and configurations
 *
 * @class Table
 */
const Table = ({ config, headers, data }) => {
  /**
   *  Renders Table headers
   *
   * @returns {JSX} JSX
   */
  const renderTableHeader = () => {
    return (
      <div className="table-row table-header">
        {headers.map(header => (
          <span
            key={header}
            className={`${config[header].width || 'table-col'}`}
          >
            {config[header].alias ? config[header].alias : header}
          </span>
        ))}
      </div>
    );
  };

  /**
   * Renders Table Body
   *
   * @returns {JSX} JSX
   */
  const renderTableBody = () => {
    const updatedData = data.map((item, index) => {
      item['serial'] = index + 1;
      return item;
    });

    return (
      <div>
        {updatedData && updatedData.map((row, id) => (
          <div className="table-row" key={id}>
            {headers.map(header => (
              <span
                className={`${config[header].width || 'table-col'}`}
                key={header}
              >
                {row[header]}
              </span>
            ))}
          </div>
        ))}
      </div>
    );
  };

  return (
    <div className="table-wrapper">
      {renderTableHeader()}
      {renderTableBody()}
    </div>
  );
};

Table.propTypes = {
  config: PropTypes.object,
  headers: PropTypes.array,
  data: PropTypes.array,
};

export default Table;
