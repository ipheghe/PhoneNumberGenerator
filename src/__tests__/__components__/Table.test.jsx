// react libraries
import React from 'react';

// third-party libraries
import { shallow } from 'enzyme';
import render from 'react-test-renderer';

// components
import Table from '../../components/Table';

const tableConfig = {
  serial: {
    alias: '#',
    width: 'table-col-20',
  },
  id: { alias: 'id', width: 'table-col-30' },
  generatedNumbers: { alias: 'Generated Numbers' },
};

const props = {
  config: tableConfig,
  data: [{ id: 1, generatedNumbers: 1234555 }],
  headers: Object.keys(tableConfig),
};

describe('Table Component', () => {
  const shallowWrapper = shallow(<Table {...props} />);

  it('should render Table component', () => {
    expect(shallowWrapper.find('.table-wrapper').length).toBe(1);
  });

  it('should match component snapshot', () => {
    const tree = render.create(<Table {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should display table header and body rows', () => {
    expect(shallowWrapper.find('.table-row').length).toBe(2);
  });

  it('should display table header', () => {
    expect(shallowWrapper.find('.table-header').length).toBe(1);
  });
});
