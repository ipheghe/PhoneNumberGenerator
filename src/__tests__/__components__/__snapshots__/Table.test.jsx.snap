// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should match component snapshot 1`] = `
<div
  className="table-wrapper"
>
  <div
    className="table-row table-header"
  >
    <span
      className="table-col-20"
    >
      #
    </span>
    <span
      className="table-col-30"
    >
      id
    </span>
    <span
      className="table-col"
    >
      Generated Numbers
    </span>
  </div>
  <div>
    <div
      className="table-row"
    >
      <span
        className="table-col-20"
      >
        1
      </span>
      <span
        className="table-col-30"
      >
        1
      </span>
      <span
        className="table-col"
      >
        1234555
      </span>
    </div>
  </div>
</div>
`;
